### DISCUSSION TITLE
Prisma Engine debugging - nodejs stack memory issue

### DISCUSSION THREAD

**Discussion by goofiw** [prisma/prisma#15971] [Q&A]
We've been having some strangely high memory usage and have identified it to be taking place in the stack, we're talking up to 12gb in a not heavily used application.  This is happening from an api server that uses prisma run in a docker container.

is there a way to debug whats going on in the prisma query engine?  It may not be the cause of our issue, but we aren't using many nodejs native modules (which use the stack vs the heap for storing objects from what I read in the node docs)

From this post a while ago, we're thinking of looking into reducing the statement_cache_size, however we do see it's already been reduced in prisma 4.0.0

https://github.com/prisma/prisma-engines/issues/2072

Another curious thing, when we inspect the running processes in our container, we don't see the prisma engine running separately like in the screenshot of that issue.  Has that changed since prisma 2.26.0?  It could also be an operating system or node version difference.  We're on nodejs 16.

**Comment by nurul3101** [comment]
Hello @goofiw 👋 

Could you try enabling [Tracing](https://www.prisma.io/docs/concepts/components/prisma-client/opentelemetry-tracing)?
Tracing should be helpful in analyzing your application's performance and identifying bottleneck. Are you seeing any performance issues in query response times? Long-running queries could potentially be an issue for high memory usage.

What is the current configuration value for `statement_cache_size`?
From Prisma 4.0 onwards the default value for `statement_cache_size` is 100 ([Reference](https://github.com/prisma/prisma/issues/7727)). I am curious to see how much performance improvement can be obtained by further reducing it.

**Reply by goofiw** [reply]
thanks @nurul3101 we'll give enabling tracing a shot and update.  We are currently using the default `statement_cache_size`

**Comment by goofiw** [comment]
@nurul3101 I read through the tracing docs, but the setup there isn't working with prisma 4.3.1 for me. (I think it may have something to do with nestjs module initialization)

I downloaded and ran the sample project, and I didn't notice any logs that would answer the question of how much memory space the engine is using.

We don't really have a performance issue, just that the memory within our docker containers seems absurdly high for the workloads.  That memory is being taken up by the stack and not the heap from using node debugging tools.  

Prisma is generating some very long queries (as in, the string size of queries are 50kb+) so I wonder if it is storing those somewhere in the stack.

Another note - we are using postgres

**Comment by janpio** [comment]
> Another curious thing, when we inspect the running processes in our container, we don't see the prisma engine running separately like in the screenshot of that issue. Has that changed since prisma 2.26.0? It could also be an operating system or node version difference. We're on nodejs 16.

Yes. In Prisma the way we run our engines changed, instead of a separate binary sidecar process it now runs as a Node-API library which is more performant. Unfortunately for your case, that means that the memory and CPU usage is not reported separately any more.

You can manually get the old behavior via setting the engine type to `binary`: https://www.prisma.io/docs/concepts/components/prisma-engines/query-engine#defining-the-query-engine-type-for-prisma-client Note that this might have slight negative effects on overall performance in some cases, so be aware. But to debug your problem this might help.

Besides that: `statement_cache_size` sounds like a red herring in your case. I think you should not waste to much time looking into that right now before really knowing what is happening.

When you can confirm via engine type `binary` that it is indeed Prisma that needs that much memory, feel free to open a bug issue here and provide all the information it asks for - then we can debug better with more information and lead you in the right direction to figure this out.

**Reply by goofiw** [reply]
Hi @janpio 

We're have high confidence it was prisma.  We lowered out connection count from 9 to 2 and the memory dropped significantly from about 12gb on some pods to ~2gb.  It is still a strange to us that the app memory usage isn't dropping once it gets up to that 2gb it will stay there and it looks like the memory usage is in the stack not the heap.

We're running our apps in docker containers on kubernetes (4cpu, 16gb nodes) and lowered the connection count after reading up on tuning for serverless environments.  Currently there are 8 separate pods.  Unfortunately our database provider doesn't currently support pgbouncer or we would have tried going that route.

Anita was nice enough to reach out on the community slack and I shared a meeting scheduling link with her.

**Reply by pantharshit00** [reply]
Hey @goofiw 

Just to make sure, did you also try `binary` engine type once? That will help us in eliminating one possible source of problem if there is change in memory usage.

**Reply by janpio** [reply]
We are talking in private now.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
